File: src/multi_agent_platform/contracts/turn_execution.py
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, model_validator

from multi_agent_platform.contracts.run_plans import PlannedTask
from multi_agent_platform.contracts.runs import TaskRecord
from multi_agent_platform.tools.registry import PlannedToolCall


class ExecutionBackend(StrEnum):
    DETERMINISTIC = "deterministic"
    LLM = "llm"


class AgentExecutionProfile(BaseModel):
    model_config = ConfigDict(extra="forbid")

    agent_name: str
    backend: ExecutionBackend = ExecutionBackend.DETERMINISTIC
    llm_provider_name: str | None = None
    model_name: str | None = None
    temperature: float | None = Field(default=None, ge=0.0, le=2.0)
    max_output_tokens: int | None = Field(default=None, ge=1)
    timeout_seconds: float | None = Field(default=None, gt=0.0)
    max_retries: int = Field(default=0, ge=0)

    @model_validator(mode="after")
    def validate_backend_settings(self) -> "AgentExecutionProfile":
        if self.backend is ExecutionBackend.DETERMINISTIC:
            if any(
                value is not None
                for value in (
                    self.llm_provider_name,
                    self.model_name,
                    self.temperature,
                    self.max_output_tokens,
                    self.timeout_seconds,
                )
            ):
                raise ValueError("Deterministic backend does not accept LLM configuration")
            return self
        if self.llm_provider_name is None:
            raise ValueError("LLM backend requires llm_provider_name")
        return self


class StructuredTurnOutput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    summary: str
    planned_tool_calls: list[PlannedToolCall] = Field(default_factory=list)


class LlmUsage(BaseModel):
    model_config = ConfigDict(extra="forbid")

    input_tokens: int = Field(default=0, ge=0)
    output_tokens: int = Field(default=0, ge=0)
    total_tokens: int = Field(default=0, ge=0)
    estimated_cost_usd: float | None = Field(default=None, ge=0.0)

    @model_validator(mode="after")
    def validate_total_tokens(self) -> "LlmUsage":
        minimum_total = self.input_tokens + self.output_tokens
        if self.total_tokens < minimum_total:
            raise ValueError("total_tokens must be at least input_tokens + output_tokens")
        return self


class LlmTurnRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    user_goal: str
    task: TaskRecord
    execution_profile: AgentExecutionProfile
    available_tool_names: list[str] = Field(default_factory=list)


class LlmPlanRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    user_goal: str
    workflow_type: str
    execution_profile: AgentExecutionProfile


class LlmTurnResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    provider_name: str
    model_name: str
    output: StructuredTurnOutput
    usage: LlmUsage = Field(default_factory=LlmUsage)
    finish_reason: str | None = None
    latency_ms: int | None = Field(default=None, ge=0)
    raw_response_text: str | None = None


class LlmExecutionOutcome(BaseModel):
    model_config = ConfigDict(extra="forbid")

    output: StructuredTurnOutput
    llm_response: LlmTurnResponse | None = None
    error_message: str | None = None
    fallback_used: bool = False
    attempt_count: int = Field(default=1, ge=1)


class LlmPlanOutput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    workflow_type: str
    template_name: str
    summary: str
    tasks: list[PlannedTask]


class LlmPlanResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    provider_name: str
    model_name: str
    output: LlmPlanOutput
    usage: LlmUsage = Field(default_factory=LlmUsage)
    finish_reason: str | None = None
    latency_ms: int | None = Field(default=None, ge=0)
    raw_response_text: str | None = None


class LlmPlanningOutcome(BaseModel):
    model_config = ConfigDict(extra="forbid")

    output: LlmPlanOutput
    llm_response: LlmPlanResponse | None = None
    error_message: str | None = None
    fallback_used: bool = False
    attempt_count: int = Field(default=1, ge=1)

File: apps/api-go/internal/contracts/turn_execution.go
package contracts

type ExecutionBackend string

const (
	ExecutionBackendDeterministic ExecutionBackend = "deterministic"
	ExecutionBackendLLM           ExecutionBackend = "llm"
)

type TaskRecord struct {
	TaskID             string   `json:"task_id"`
	Title              string   `json:"title"`
	Description        string   `json:"description"`
	AssignedAgent      string   `json:"assigned_agent"`
	Status             string   `json:"status"`
	DependencyIDs      []string `json:"dependency_ids"`
	AcceptanceCriteria []string `json:"acceptance_criteria"`
	RiskLevel          string   `json:"risk_level"`
	AttemptCount       int      `json:"attempt_count"`
}

type AgentExecutionProfile struct {
	AgentName       string           `json:"agent_name"`
	Backend         ExecutionBackend `json:"backend"`
	LLMProviderName *string          `json:"llm_provider_name,omitempty"`
	ModelName       *string          `json:"model_name,omitempty"`
	Temperature     *float64         `json:"temperature,omitempty"`
	MaxOutputTokens *int             `json:"max_output_tokens,omitempty"`
	TimeoutSeconds  *float64         `json:"timeout_seconds,omitempty"`
	MaxRetries      int              `json:"max_retries"`
}

type LLMUsage struct {
	InputTokens      int      `json:"input_tokens"`
	OutputTokens     int      `json:"output_tokens"`
	TotalTokens      int      `json:"total_tokens"`
	EstimatedCostUSD *float64 `json:"estimated_cost_usd,omitempty"`
}

type PlannedToolCall struct {
	ToolName  string            `json:"tool_name"`
	ToolInput map[string]string `json:"tool_input"`
}

type StructuredTurnOutput struct {
	Summary          string            `json:"summary"`
	PlannedToolCalls []PlannedToolCall `json:"planned_tool_calls"`
}

type LLMWorkerTurnRequest struct {
	RunID              string                `json:"run_id"`
	UserGoal           string                `json:"user_goal"`
	Task               TaskRecord            `json:"task"`
	ExecutionProfile   AgentExecutionProfile `json:"execution_profile"`
	AvailableToolNames []string              `json:"available_tool_names"`
}

type LLMWorkerPlanRequest struct {
	RunID            string                `json:"run_id"`
	UserGoal         string                `json:"user_goal"`
	WorkflowType     string                `json:"workflow_type"`
	ExecutionProfile AgentExecutionProfile `json:"execution_profile"`
}

type LLMTurnResponse struct {
	ProviderName    string               `json:"provider_name"`
	ModelName       string               `json:"model_name"`
	Output          StructuredTurnOutput `json:"output"`
	Usage           LLMUsage             `json:"usage"`
	FinishReason    *string              `json:"finish_reason,omitempty"`
	LatencyMS       *int                 `json:"latency_ms,omitempty"`
	RawResponseText *string              `json:"raw_response_text,omitempty"`
}

type LLMExecutionOutcome struct {
	Output        StructuredTurnOutput `json:"output"`
	LLMResponse   *LLMTurnResponse     `json:"llm_response,omitempty"`
	ErrorMessage  *string              `json:"error_message,omitempty"`
	FallbackUsed  bool                 `json:"fallback_used"`
	AttemptCount  int                  `json:"attempt_count"`
}

type PlannedTask struct {
	TaskID             string   `json:"task_id"`
	Title              string   `json:"title"`
	Description        string   `json:"description"`
	AssignedAgent      string   `json:"assigned_agent"`
	DependencyIDs      []string `json:"dependency_ids"`
	AcceptanceCriteria []string `json:"acceptance_criteria"`
}

type LLMPlanReport struct {
	RunID        string        `json:"run_id"`
	WorkflowType string        `json:"workflow_type"`
	TemplateName string        `json:"template_name"`
	Summary      string        `json:"summary"`
	Tasks        []PlannedTask `json:"tasks"`
}

type LLMPlanResponse struct {
	ProviderName    string        `json:"provider_name"`
	ModelName       string        `json:"model_name"`
	Output          LLMPlanReport `json:"output"`
	Usage           LLMUsage      `json:"usage"`
	FinishReason    *string       `json:"finish_reason,omitempty"`
	LatencyMS       *int          `json:"latency_ms,omitempty"`
	RawResponseText *string       `json:"raw_response_text,omitempty"`
}

type LLMPlanningOutcome struct {
	Output       LLMPlanReport    `json:"output"`
	LLMResponse  *LLMPlanResponse `json:"llm_response,omitempty"`
	ErrorMessage *string          `json:"error_message,omitempty"`
	FallbackUsed bool             `json:"fallback_used"`
	AttemptCount int              `json:"attempt_count"`
}
